Add voting-only master node

ywelsch opened the pull request:
A voting-only master-eligible node is a node that can participate in master elections but will not act as a master in the cluster. In particular, a voting-only node can help elect another master-eligible node as master, and can serve as a tiebreaker in elections. High availability (HA) clusters require at least three master-eligible nodes, so that if one of the three nodes is down, then the remaining two can still elect a master amongst them-selves. This only requires one of the two remaining nodes to have the capability to act as master, but both need to have voting powers. This means that one of the three master-eligible nodes can be made as *voting-only*. If this voting-only node is a dedicated master, a less powerful machine or a smaller heap-size can be chosen for this node. Alternatively, a voting-only non-dedicated master node can play the role of the third master-eligible node, which allows running an HA cluster with only two dedicated master nodes.

Closes #14340

This is joint work with @DaveCTurner

DaveCTurner committed:
Voting-only nodes

Adds a plugin that allows some master-eligible nodes to be used as
"voting-only" nodes. Such nodes are never elected as the true master of the
cluster, but they perform all the tasks needed of a master-eligible node to
provide resilience.

DaveCTurner committed:
Randomise good-quorum calculation in CoordinationStateTests

ywelsch committed:
State transfer only

ywelsch committed:
Move to isElectionQuorum

ywelsch committed:
use JoinVoteCollection

ywelsch committed:
fix test

ywelsch committed:
fix build and use transport intercepter

ywelsch committed:
move tests

ywelsch committed:
move tests

ywelsch committed:
rest test

ywelsch committed:
Add x-pack feature set

ywelsch committed:
add docs

ywelsch committed:
fixup

ywelsch committed:
fix docs tests

ywelsch committed:
Register usage action

ywelsch committed:
more fixups

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into state-transfer-only

ywelsch committed:
more fixes

ywelsch committed:
More fixups

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into state-transfer-only

ywelsch committed:
fix docs tests on OSS distrib

jasontedor commented on docs/reference/cluster.asciidoc:
```diff
@@ -22,12 +22,14 @@ one of the following:
+* `master:false`, `data:false`, `ingest:false`, `voting_only:true`, or
+  `coordinating_only:false`, which respectively remove from the subset all
+  master-eligible nodes, all data nodes, all ingest nodes, all voting-only
+  nodes and all coordinating-only nodes.
```
I think the docs should clearly delineate that `voting_only` requires x-pack, we can wrap it in conditionals and add x-pack annotations so that it doesn't show in the docs if someone builds the OSS-only docs, and has x-pack designations in the docs when the full docs are published.

ywelsch replied:
I didn't know that OSS-only docs were even a thing. Are we publishing those somewhere? How do you build those? You will have to spell out the details on how to set up the conditionals, I'm not aware of any such infrastructure.

jrodewig replied:
While we do have `[x-pack]` macros, I'm not aware of an OSS-only docs build functionality.

@debadair @lcawl Are you aware of an OSS-only docs build?

jasontedor replied:
We don’t build them but it’s available for users that want to build OSS-only docs.

@lcawl Can you help @ywelsch add the appropriate x-pack annotations here?

debadair replied:
Content that requires the default distribution should be tagged with the [role="xpack"] directive. Now that all of the doc source is in the public repo we no longer maintain two versions of the index.asciidoc file, so conditional statements based on the `include_xpack` attribute have no effect. 

Inline references like this are tricky. If using the `voting_only` attribute throws an error in the OSS distro, I'd be inclined to add a note to that effect. Something like:

> NOTE: Designating nodes as `voting_only` and using `voting_only` in node filters is requires the default distribution of Elasticsearch.

@lcawl can correct me if I'm wrong, but I don't think there's (currently) any way to attach the xpack bug to an admonition block.

DaveCTurner commented on docs/reference/modules/node.asciidoc:
```diff
@@ -134,6 +134,37 @@ cluster.remote.connect: false <4>
+machine or a smaller heap-size can be chosen for this node. Alternatively,
+a voting-only non-dedicated master node can play the role of the third
+master-eligible node, which allows running an HA cluster with only two
+dedicated master nodes.
```
I think we should make it clearer that voting-only nodes still need a fast disk and a low-latency connection to the true masters in order to be effective, and will use the same amount of disk space as any other dedicated master node. They should need less CPU and less heap.

ywelsch replied:
I've tried to address this in ccdb483eaeb. Let me know if that's ok, or perhaps suggest an alternative wording.

DaveCTurner reviewed (commented):
I have left a few minor comments but this all looks good.

I have read, but not properly reviewed, all the code that's needed to integrate this plugin with X-pack. That's not something with which I'm familiar enough to pass an opinion.

ywelsch committed:
Fold JoinVoteCollection into VoteCollection

rjernst commented on server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java:
```diff
@@ -0,0 +1,49 @@
+
+/**
+ * Allows plugging in a custom election strategy. Note that, in order to guarantee safety of the system, custom election strategies should
+ * only be more restrictive than the behavior that's provided by {@link DefaultElectionStrategy}.
```
Could we enforce this restriction? For example, instead of having this as an interface, have an abstract class, with a final method that is public, and the overrideability is a protected method that only adds to the restrictions the main final method enforces.

ywelsch replied:
Good idea. Adapted in e284426.

ywelsch committed:
s/election type/election strategy/

rjernst commented on server/src/main/java/org/elasticsearch/plugins/DiscoveryPlugin.java:
```diff
@@ -85,4 +86,8 @@
      */
     default BiConsumer<DiscoveryNode,ClusterState> getJoinValidator() { return null; }
+
+    default Map<String, ElectionStrategy> getElectionStrategies() {
```
Please add at least a basic javadoc. I would reference here the setting that is used to set the election strategy.

ywelsch replied:
fixed in e284426

ywelsch committed:
test adjustment

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java:
```diff
@@ -117,6 +117,11 @@ private XPackSettings() {
 
+    /** Setting for enabling or disabling the voting-only-node functionality. Needs to be enabled on both voting-only nodes and regular
+     *  master-eligible nodes for the voting-only functionality to work correctly. Defaults to true. */
+    public static final Setting<Boolean> VOTING_ONLY_ENABLED = Setting.boolSetting("xpack.voting_only.enabled", true,
```
Why is it necessary to have a separate enabling of the feature? Why coudln't it just always be available with xpack (especially since you have the default as enabled)? Why would someone need to disable this feature, when they would just not set any nodes to voting only?

ywelsch replied:
Good question. My thinking was the following: We have this setting for all other features as well. When the setting is disabled, the other plugins are not adding anything to the extension points. I thought this could be used as an extra safeguard that would allow completely disabling the functionality provided by this plugin in case something was wrong with it, given that it's always bundled with x-pack. I'm fine not having this setting if you think that that is not necessary. I did not see a strong reason to deviate from the norm here.

rjernst replied:
IMO it contributes to an overload of settings users could mess with, with no real added benefit. I see what you mean about all the other xpack plugins. I wonder if we really need that though. For eg security it makes sense, but why would one ever disable deprecation, or rollup, or ilm? These features do not actively do anything unless used, afaik. For features like that, I think we should just have it always "enabled", but I'm open to hearing others arguments for why we need the ability to disable across all xpack features.

ywelsch replied:
For features that provide API endpoints, it makes perhaps more sense (so that these API endpoints don't even become available). For voting_only nodes, the only thing that becomes visible is that the `voting_only` role appears in several stats endpoints, even if there is no voting_only node in the cluster. The node counts, that are part of the cluster stats, for example, expose this info.
I don't hold a particularly strong opinion either way on this, so I've removed the setting in bcd6ec2. I can revert that if others feel differently about this.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -539,4 +544,24 @@ public int hashCode() {
+    /**
+     * A collection of votes, extending {@link VoteCollection}, which additionally records the Joins
+     */
+    public static class JoinVoteCollection extends VoteCollection {
```
fixed in e7c325e

ywelsch commented on server/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java:
```diff
@@ -75,6 +76,11 @@
+    public static final String DEFAULT_ELECTION_TYPE = "default";
+
+    public static final Setting<String> ELECTION_TYPE_SETTING =
+        new Setting<>("cluster.election.type", DEFAULT_ELECTION_TYPE, Function.identity(), Property.NodeScope);
```
fixed in 1caa1b6

ywelsch commented on x-pack/plugin/voting-only-node/src/test/java/org/elasticsearch/cluster/coordination/VotingOnlyNodePluginTests.java:
```diff
@@ -0,0 +1,105 @@
+
+        // start a fresh full master node, which will be brought into the cluster as master by the voting-only nodes
+        final String newMaster = internalCluster().startNode();
+        assertEquals(newMaster, internalCluster().getMasterName());
```
fixed in 6907d06

ywelsch commented on x-pack/plugin/voting-only-node/src/test/java/org/elasticsearch/cluster/coordination/VotingOnlyNodePluginTests.java:
```diff
@@ -0,0 +1,105 @@
+            assertThat(client().admin().cluster().prepareState().setMasterNodeTimeout("100ms")
+                .execute().actionGet().getState().nodes().getMasterNodeId(), nullValue());
+            fail("should not be able to find master");
+        } catch (MasterNotDiscoveredException e) {
```
fixed in 6907d06

rjernst reviewed (commented):
I have a couple comments on the plugin api.

ywelsch committed:
Move ElectionStrategy from interface to class

ywelsch committed:
Have VotingOnlyNodePlugin always enabled

rjernst commented on server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java:
```diff
@@ -0,0 +1,53 @@
+                lastAcceptedConfiguration, joinVotes);
+    }
+
+    protected boolean isCustomElectionQuorum(DiscoveryNode localNode, long localCurrentTerm, long localAcceptedTerm,
```
javadocs on this would be good, explaining what a plugin author extending this should return. I would also reference the final method, explaining how the results are merged together. I might also name this with terminology like "additional" since it adds to the real quorum method.

ywelsch replied:
fixed in 264e5a3

rjernst commented on server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Allows plugging in a custom election strategy, restricting the notion of an election quorum.
+ */
+public class ElectionStrategy {
```
why not make this abstract and the protected method abstract? The default instance can be an anonymous class returning true. This way someone extending it does not eg mispell the method, at the same time forgetting @Override, and not understanding why their method is not being called.

ywelsch replied:
good idea. adapted in 264e5a3

ywelsch committed:
Ryan feedback

ywelsch committed:
checkstyle

ywelsch committed:
doc changes

ywelsch committed:
Add note about voting-only in default distrib

ywelsch commented:
@elasticmachine run elasticsearch-ci/docs-check

rjernst reviewed (approved):
LGTM from the plugin side

DaveCTurner committed:
Reword docs

* Generalise the note on the node filters page to talk about the `voting_only`
  role, since it's not clear on this page what "designating" means.

* Contrast `default-dist` to `oss-dist`, because in isolation the phrase
  "default distribution" is kinda mysterious to the general public.

* Qualify statement about "any master eligible node may be elected"

* Add excuse for the confusing terminology

* Shrink explanation about HA clusters, because it involved concepts like
  "voting powers" and "capability to act as master" that aren't really defined.

* Remove suggestion to use a data node as a voting-only node: although this
  might work in many cases, I think if we specifically mention it here we will
  see too many clusters with overloaded data nodes that can't reasonably be a
  master. I think this is best left implicit.

DaveCTurner reviewed (approved):
I pushed 7bb6fa28 to rework the docs a bit, which could do with a review from someone that isn't me. Everything else LGTM.

ywelsch commented:
build succeeded, but Github hook failed, so I'm considering this as green to merge.

> 14:15:05 BUILD SUCCESSFUL in 44m 27s
> 14:15:05 2504 actionable tasks: 2278 executed, 226 up-to-date
> 14:16:39 Setting status of 7bb6fa28654e4965bb44d5ddaf33fbd3320f8b98 to SUCCESS with url https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/1385/ and message: 'Build finished. '
14:16:39 Using context: elasticsearch-ci/1
14:16:39 Could not update commit status of the Pull Request on GitHub.
14:16:39 org.kohsuke.github.HttpException: Server returned HTTP response code: 500, message: 'Internal Server Error' for URL: https://api.github.com/repos/elastic/elasticsearch/statuses/7bb6fa28654e4965bb44d5ddaf33fbd3320f8b98
14:16:39 	at org.kohsuke.github.Requester.parse(Requester.java:646)
14:16:39 	at org.kohsuke.github.Requester.parse(Requester.java:607)
14:16:39 	at org.kohsuke.github.Requester._to(Requester.java:285)
14:16:39 	at org.kohsuke.github.Requester.to(Requester.java:247)
14:16:39 	at org.kohsuke.github.GHRepository.createCommitStatus(GHRepository.java:1104)
14:16:39 	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.createCommitStatus(GhprbSimpleStatus.java:283)
14:16:39 	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.onBuildComplete(GhprbSimpleStatus.java:241)
14:16:39 	at org.jenkinsci.plugins.ghprb.GhprbBuilds.onCompleted(GhprbBuilds.java:205)
14:16:39 	at org.jenkinsci.plugins.ghprb.GhprbBuildListener.onCompleted(GhprbBuildListener.java:28)
14:16:39 	at hudson.model.listeners.RunListener.fireCompleted(RunListener.java:209)
14:16:39 	at hudson.model.Run.execute(Run.java:1863)
14:16:39 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
14:16:39 	at hudson.model.ResourceController.execute(ResourceController.java:97)
14:16:39 	at hudson.model.Executor.run(Executor.java:429)
14:16:39 Caused by: java.io.IOException: Server returned HTTP response code: 500 for URL: https://api.github.com/repos/elastic/elasticsearch/statuses/7bb6fa28654e4965bb44d5ddaf33fbd3320f8b98
14:16:39 	at sun.reflect.GeneratedConstructorAccessor308.newInstance(Unknown Source)
14:16:39 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
14:16:39 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
14:16:39 	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1944)
14:16:39 	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1939)
14:16:39 	at java.security.AccessController.doPrivileged(Native Method)
14:16:39 	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1938)
14:16:39 	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1508)
14:16:39 	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
14:16:39 	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:263)
14:16:39 	at org.kohsuke.github.Requester.parse(Requester.java:625)
14:16:39 	... 13 more
